[Feature branch] Add Neural Stats API

q-andy opened the pull request:
### Description
Implementing Neural Stats API framework design proposed in #1196. This initial PR sets up the foundation for the framework to track event and state stats throughout the neural search plugin and exposed their values via API.

![Image](https://github.com/user-attachments/assets/3ab87a78-a913-4349-b10e-fc7717623f82)

* **Event-based stats**
    * Event stats are recorded in code at a node level (processor executions, documents ingested, etc)
    * When an API call is made, all node-level maps are fetched via transport action and returned in the response.
* **State stats**
    * State stats are defined by helper functions that populate state stat values
    * When an API call is made, the functions are invoked and the information is added to the response on demand

See RFC for more details.

Initial implementation includes 3 stats:
- Text embedding processor executions
- Text embedding processors in pipelines
- Cluster version

### Related Issues
Resolves #1196 #1104 
Related: #1146 

### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

### Example requests


```
GET /_plugins/_neural/stats
{
	"_nodes": {
		"total": 1,
		"successful": 1,
		"failed": 0
	},
	"cluster_name": "integTest",
	"cluster_version": "3.0.0",
	"processors": {
		"ingest": {
			"text_embedding_processors_in_pipelines": 0
		}
	},
	"all_nodes": {
		"processors": {
			"ingest": {
				"text_embedding_executions": 0
			}
		}
	},
	"nodes": {
		"r-hVPa7-Ra6FBbdpfVrHjg": {
			"processors": {
				"ingest": {
					"text_embedding_executions": 0
				}
			}
		}
	}
}
```

```
GET /_plugins/_neural/stats?include_metadata=true
{
	"_nodes": {
		"total": 1,
		"successful": 1,
		"failed": 0
	},
	"cluster_name": "integTest",
	"cluster_version": {
		"value": "3.0.0",
		"stat_type": "settable"
	},
	"processors": {
		"ingest": {
			"text_embedding_processors_in_pipelines": {
				"value": 0,
				"stat_type": "countable"
			}
		}
	},
	"all_nodes": {
		"processors": {
			"ingest": {
				"text_embedding_executions": {
					"value": 0,
					"stat_type": "timestamped_counter",
					"trailing_interval_value": 0,
					"minutes_since_last_event": 29018783
				}
			}
		}
	},
	"nodes": {
		"r-hVPa7-Ra6FBbdpfVrHjg": {
			"processors": {
				"ingest": {
					"text_embedding_executions": {
						"value": 0,
						"stat_type": "timestamped_counter",
						"trailing_interval_value": 0,
						"minutes_since_last_event": 29018783
					}
				}
			}
		}
	}
}
```

```
GET _plugins/_neural/stats/text_embedding_executions?include_metadata=true&flat_keys=true

{
	"_nodes": {
		"total": 1,
		"successful": 1,
		"failed": 0
	},
	"cluster_name": "integTest",
	"all_nodes.processors.ingest.text_embedding_executions": {
		"value": 0,
		"stat_type": "timestamped_counter",
		"trailing_interval_value": 0,
		"minutes_since_last_event": 29018784
	},
	"nodes": {
		"r-hVPa7-Ra6FBbdpfVrHjg": {
			"processors.ingest.text_embedding_executions": {
				"value": 0,
				"stat_type": "timestamped_counter",
				"trailing_interval_value": 0,
				"minutes_since_last_event": 29018784
			}
		}
	}
}
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/neural-search/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

heemin32 commented:
Should it be full flat?
```
GET _plugins/_neural/stats/text_embedding_executions?include_metadata=true&flat_keys=true

{
	"_nodes.total": 1,
        "_nodes.successful": 1,
        "_nodes.failed": 0
        "all_nodes.processors.ingest.text_embedding_executions.value": 0,
        ...
        "nodes.r-hVPa7-Ra6FBbdpfVrHjg.processors.ingest.text_embedding_executions.minutes_since_last_event": 29018784
}
```

q-andy commented:
It can go either way, but I think preserving some basic structure is helpful for navigating the high level categories. For example, cluster setting still divides the flat keys between `persistent`, `transient`, and `defaults` fields.

The way I see it, flatten is an option to simplify navigating the nested fields, not a hard rule to flatten everything, and keeping some info like metadata grouped together makes it easier to parse since than if it was completely flat.

Example cluster settings call with flatten enabled:

```
GET _cluster/settings?include_defaults=true&flat_settings=true

{
	"persistent": {
		"plugins.ml_commons.model_access_control_enabled": "true",
		"plugins.ml_commons.native_memory_threshold": "99",
		"plugins.ml_commons.only_run_on_ml_node": "false",
		"plugins.neural_search.stats_enabled": "true"
	},
	"transient": {},
	"defaults": {
		"_plugin_neural_search_hybrid_query_executor.queue_size": "1000",
		"_plugin_neural_search_hybrid_query_executor.size": "20",
		"action.auto_create_index": "true",
         . . .
       }
}
```
Compare to non-flat

```
GET _cluster/settings?include_defaults=true
{
	"persistent": {
		"plugins": {
			"neural_search": {
				"stats_enabled": "true"
			},
			"ml_commons": {
				"only_run_on_ml_node": "false",
				"model_access_control_enabled": "true",
				"native_memory_threshold": "99"
			}
		}
	},
	"transient": {},
	"defaults": {
		"task_resource_tracking": {
			"enabled": "true"
		},
	. . .
        },
        . . . 
}

heemin32 commented:
That make sense. We only need to flatten the key for easier search. Then, shouldn't it be like this?
```
"all_nodes" {
      "processors.ingest.text_embedding_executions": {
		"value": 0,
		"stat_type": "timestamped_counter",
		"trailing_interval_value": 0,
		"minutes_since_last_event": 29018784
	},
}
```

Also wondering if this could be better.
```
"all_nodes" {
      "processors.ingest.text_embedding_executions.value": 0,
      "processors.ingest.text_embedding_executions.stat_type": "timestamped_counter",
      "processors.ingest.text_embedding_executions.trailing_interval_value": 0,
      "processors.ingest.text_embedding_executions.minutes_since_last_event": 29018784
}
```

heemin32 commented:
There are two components: Stats API and Metric Management. For the Metric Management code, it might be better to have a separate package to make it more portable for use in other plugins.

q-andy commented:
> Then, shouldn't it be like this? (all_nodes unflattened)

Yes, that makes more sense, good catch

> Also wondering if this could be better. (metadata flattened)

I considered that as well, what I'm thinking is the "stat metadata" object replaces the raw value so the flat key name is the same whether or not metadata is included or not. Like the format will always be at `"path.to.stat: <data>"`, sometimes the data is the raw value, sometimes its a the metadata, but the stat key name is same.

It also keeps things organized when you parse recursively. For example. if a cluster manager is periodically calling the stats API and parsing the response without hardcoding stat names, if we keep metadata non-flat, for every stat you can foreach on the categories with flat_keys and parse each stat and all its metadata one by one in a single pass. Whereas if it was completely flat metadata, running a foreach would need additional logic to split, sort, and group the metadata keys by the stat name again. Also you would probably never be interested in the value of `path.to.stat.stat_type` independently of the other stat metadata.

Just some thoughts. I don't see the usecase right now but I could still include an option to flatten the stat metadata as well, it's a pretty small change.

q-andy commented:
Changed base to merge to feature branch instead of main

vibrantvarun commented:
Please add the change in changelog.md

vibrantvarun commented on src/main/java/org/opensearch/neuralsearch/processor/util/RestActionUtils.java:
```diff
@@ -0,0 +1,19 @@
+
+import java.util.Optional;
+
+public class RestActionUtils {
```
Take reference from ProcessorUtils

vibrantvarun commented on src/main/java/org/opensearch/neuralsearch/rest/RestNeuralStatsHandler.java:
```diff
@@ -0,0 +1,157 @@
+import static org.opensearch.neuralsearch.processor.util.RestActionUtils.splitCommaSeparatedParam;
+
+@Log4j2
+@AllArgsConstructor
```
Add javadoc

vibrantvarun commented on src/main/java/org/opensearch/neuralsearch/rest/RestNeuralStatsHandler.java:
```diff
@@ -0,0 +1,157 @@
+        .map(String::toLowerCase)
+        .collect(Collectors.toSet());
+
+    private static final Set<String> STATE_STAT_NAMES = EnumSet.allOf(StateStatName.class)
```
Can you come up with more meaninful name? State_stat is not that reader friendly.

q-andy replied:
Sure, I'll rename to `InfoStat`

vibrantvarun commented on src/main/java/org/opensearch/neuralsearch/rest/RestNeuralStatsHandler.java:
```diff
@@ -0,0 +1,157 @@
+
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) {
+        if (settingsAccessor.getIsStatsEnabled() == false) {
+            // Process params, or else will automatically return a 400 instead of a 403
+            splitCommaSeparatedParam(request, "nodeId");
+            splitCommaSeparatedParam(request, "stat");
+            request.paramAsBoolean(FLATTEN_PARAM, false);
+            request.paramAsBoolean(INCLUDE_METADATA_PARAM, false);
+
+            return channel -> channel.sendResponse(new BytesRestResponse(RestStatus.FORBIDDEN, "Stats endpoint is disabled"));
+        }
```
Move this to a separate dedicated method as validation method.

QQ: why do you need to call split methods and updating request if you anyway sending forbidden request exception at line 80.

vibrantvarun commented on src/main/java/org/opensearch/neuralsearch/processor/util/RestActionUtils.java:
```diff
@@ -0,0 +1,19 @@
+
+import java.util.Optional;
+
+public class RestActionUtils {
```
```suggestion
class RestActionUtils {
```

Make this class package-private

vibrantvarun commented on src/main/java/org/opensearch/neuralsearch/processor/util/RestActionUtils.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.Optional;
+
+public class RestActionUtils {
+    public static Optional<String[]> splitCommaSeparatedParam(RestRequest request, String paramName) {
```
Why we need this class and validations seperately. Similar kind of methods are already being used by opensearch-core. Why not reuse them? Check RestSearchAction in OpenSearch core for the same.

vibrantvarun commented on src/main/java/org/opensearch/neuralsearch/rest/RestNeuralStatsHandler.java:
```diff
@@ -0,0 +1,157 @@
+     * @param request Rest request
+     * @return NeuralStatsRequest
+     */
+    private NeuralStatsRequest getRequest(RestRequest request) {
```
```suggestion
    private NeuralStatsRequest createNeuralStatsRequest(RestRequest request) {
```

vibrantvarun commented on src/main/java/org/opensearch/neuralsearch/rest/RestNeuralStatsHandler.java:
```diff
@@ -0,0 +1,157 @@
+        return neuralStatsRequest;
+    }
+
+    NeuralStatsInput createNeuralStatsInputFromRequestParams(RestRequest request) {
```
```suggestion
    private NeuralStatsInput createNeuralStatsInputFromRequestParams(RestRequest request) {
```

vibrantvarun commented on src/main/java/org/opensearch/neuralsearch/rest/RestNeuralStatsHandler.java:
```diff
@@ -0,0 +1,157 @@
+        neuralStatsInput.setIncludeMetadata(includeMetadata);
+
+        // Determine which stat names to retrieve based on user parameters
+        Optional<String[]> stats = splitCommaSeparatedParam(request, "stat");
+        boolean retrieveAllStats = true;
+
+        // Add stats to input to retrieve if specified
+        if (stats.isPresent()) {
+            for (String stat : stats.get()) {
+                stat = stat.toLowerCase(Locale.ROOT);
+
+                if (EVENT_STAT_NAMES.contains(stat)) {
+                    retrieveAllStats = false;
+                    neuralStatsInput.getEventStatNames().add(EventStatName.from(stat));
+                } else if (STATE_STAT_NAMES.contains(stat)) {
+                    retrieveAllStats = false;
+                    neuralStatsInput.getStateStatNames().add(StateStatName.from(stat));
+                }
+            }
+        }
+
+        // If no stats are specified, add all stats to retrieve all by default
+        if (retrieveAllStats) {
+            neuralStatsInput.getEventStatNames().addAll(EnumSet.allOf(EventStatName.class));
+            neuralStatsInput.getStateStatNames().addAll(EnumSet.allOf(StateStatName.class));
+        }
+        return neuralStatsInput;
```
```suggestion
  Optional<String[]> stats = splitCommaSeparatedParam(request, "stat");
    
    if (stats.isPresent()==false) {
        // No specific stats requested, add all stats by default
        addAllStats(neuralStatsInput);
        return;
    }

    // Process requested stats
    boolean anyStatAdded = processRequestedStats(stats.get(), neuralStatsInput);
    
    // If no valid stats were added, fall back to all stats
    if (anyStatAdded ==false) {
        addAllStats(neuralStatsInput);
    }
    
    return neuralStatsInput;
  }

private boolean processRequestedStats(String[] stats, NeuralStatsInput neuralStatsInput) {
    boolean statAdded = false;
    
    for (String stat : stats) {
        String normalizedStat = stat.toLowerCase(Locale.ROOT);
        
        if (EVENT_STAT_NAMES.contains(normalizedStat)) {
            neuralStatsInput.getEventStatNames().add(EventStatName.from(normalizedStat));
            statAdded = true;
        } else if (STATE_STAT_NAMES.contains(normalizedStat)) {
            neuralStatsInput.getStateStatNames().add(StateStatName.from(normalizedStat));
            statAdded = true;
        }
    }
    
    return statAdded;
}

private void addAllStats(NeuralStatsInput neuralStatsInput) {
    neuralStatsInput.getEventStatNames().addAll(EnumSet.allOf(EventStatName.class));
    neuralStatsInput.getStateStatNames().addAll(EnumSet.allOf(StateStatName.class));
}
```

vibrantvarun replied:
Clean code.

vibrantvarun commented on src/main/java/org/opensearch/neuralsearch/settings/NeuralSearchSettings.java:
```diff
@@ -34,4 +34,11 @@ public final class NeuralSearchSettings {
         Setting.Property.NodeScope
     );
+
+    public static final Setting<Boolean> NEURAL_STATS_ENABLED = Setting.boolSetting(
```
Add javadoc

vibrantvarun commented on src/main/java/org/opensearch/neuralsearch/settings/NeuralSearchSettingsAccessor.java:
```diff
@@ -0,0 +1,44 @@
+import org.opensearch.common.settings.Settings;
+import org.opensearch.neuralsearch.stats.events.EventStatsManager;
+
+public class NeuralSearchSettingsAccessor {
```
Add javadoc

vibrantvarun commented on src/main/java/org/opensearch/neuralsearch/rest/RestNeuralStatsHandler.java:
```diff
@@ -0,0 +1,157 @@
+        .map(String::toLowerCase)
+        .collect(Collectors.toSet());
+
+    private NeuralSearchSettingsAccessor settingsAccessor;
```
Move this declaration to top

vibrantvarun commented on src/main/java/org/opensearch/neuralsearch/settings/NeuralSearchSettingsAccessor.java:
```diff
@@ -0,0 +1,44 @@
+    private boolean initialized;
+
+    @Getter
+    private volatile Boolean isStatsEnabled;
```
```suggestion
    private volatile boolean isStatsEnabled;
```

vibrantvarun commented on src/main/java/org/opensearch/neuralsearch/stats/NeuralStatsInput.java:
```diff
@@ -0,0 +1,168 @@
+import java.util.Set;
+
+/**
+ * Holds user requester input parameters for retrieving neural stats via rest handler API.
+ * It allows filtering statistics by node IDs, event statistic types, and state statistic types.
```
```suggestion
 * Entity class that holds input parameters for retrieving neural stats.
 * It is responsible for filtering statistics by node IDs, event statistic types, and state statistic types.
```

vibrantvarun commented on src/main/java/org/opensearch/neuralsearch/stats/events/EventStatsManager.java:
```diff
@@ -0,0 +1,107 @@
+import java.util.Map;
+
+/**
+ * Manager class for event counter stats, used to track monotonically increasing counts plus some possible metadata
```
```suggestion
 * Manager class for event counter stats, used to track monotonically increasing counts plus some possible metadata
```

Please rewrite to make it more reader friendly. It should provide high level information of this class.

vibrantvarun reviewed (commented):
Completed 1st round of review.

heemin32 commented on src/main/java/org/opensearch/neuralsearch/stats/NeuralStatsInput.java:
```diff
@@ -0,0 +1,168 @@
+     * @throws IOException if there's an error reading from the stream
+     */
+    public NeuralStatsInput(StreamInput input) throws IOException {
+        nodeIds = input.readBoolean() ? new HashSet<>(input.readStringList()) : new HashSet<>();
```
`readOptionalStringList();` might be better in terms of readability.
Wonder if we can make nodeIds as List so that we can avoid this list to set conversion.

heemin32 commented on src/main/java/org/opensearch/neuralsearch/stats/NeuralStatsInput.java:
```diff
@@ -0,0 +1,168 @@
+     * @param set the EnumSet to write
+     * @throws IOException If there's an error writing to the stream
+     */
+    private void writeOptionalEnumSet(StreamOutput out, EnumSet<?> set) throws IOException {
```
Might be good to have corresponding `readOptionalEnumSet()`
Better if we can add this method inside core :)

heemin32 replied:
Until core is ready, we can have the method in this class temporary.

heemin32 commented on src/main/java/org/opensearch/neuralsearch/stats/events/EventStatsManager.java:
```diff
@@ -0,0 +1,107 @@
+     */
+    public void inc(EventStatName eventStatName) {
+        boolean enabled = settingsAccessor.getIsStatsEnabled();
+        if (enabled && stats.containsKey(eventStatName)) {
```
This could become an issue when there are many `eventStatName` values. While `HashMap.get` is generally constant time, its performance may degrade as the number of items grows. Given that this method is called frequently, we may need to explore an alternative approach to access the instance directly for better efficiency.

q-andy replied:
Sure, I can have the EventStat be a member of the EventStatName enum and be instantiated in the enum constructor. That way we can access it directly without using a map at all.

If we have performance issues with hash map get, though, we do have other options, my understanding is the O(n) worst case happens when there are key object hash code collisions leading to putting objects in the same bucket and requiring a search on the bucket. There are ways we can mitigate this like implementing our hash code methods to ensure uniqueness, and Java has its own optimizations to address this like [storing collisions in a tree rather than linked list](https://openjdk.org/jeps/180).

I'll update again once I finish running performance benchmarks

heemin32 replied:
Yes. Storing it on a tree is also not O(1). Also, even if the hashcode are distinct, I don't think they will all have their own bucket. The number of bucket can be smaller then the number of hashcode.

heemin32 reviewed (commented):
Overall, I have two comments:  

1. The most frequently called method (currently the `increment` method) should always operate in O(1) time complexity, regardless of the number of metrics defined in the future. `Hash.get()` does not always guarantee O(1) performance when a large number of items are stored.  

2. We should avoid using the `public static instance()` pattern. This approach should only be considered as a last resort when no other alternatives exist, as it can make unit testing more difficult.

heemin32 commented:
> > Then, shouldn't it be like this? (all_nodes unflattened)
> 
> Yes, that makes more sense, good catch
> 
> > Also wondering if this could be better. (metadata flattened)
> 
> I considered that as well, what I'm thinking is the "stat metadata" object replaces the raw value so the flat key name is the same whether or not metadata is included or not. Like the format will always be at `"path.to.stat: <data>"`, sometimes the data is the raw value, sometimes its a the metadata, but the stat key name is same.
> 
> It also keeps things organized when you parse recursively. For example. if a cluster manager is periodically calling the stats API and parsing the response without hardcoding stat names, if we keep metadata non-flat, for every stat you can foreach on the categories with flat_keys and parse each stat and all its metadata one by one in a single pass. Whereas if it was completely flat metadata, running a foreach would need additional logic to split, sort, and group the metadata keys by the stat name again. Also you would probably never be interested in the value of `path.to.stat.stat_type` independently of the other stat metadata.
> 
> Just some thoughts. I don't see the usecase right now but I could still include an option to flatten the stat metadata as well, it's a pretty small change.

Yes. One use case that I can think of is that, when operator want to grep stats containing processor, this whole flat option can be handy.
```
_plugin/neural/stats?flat=true | grep processor
```

q-andy commented on src/main/java/org/opensearch/neuralsearch/settings/NeuralSearchSettingsAccessor.java:
```diff
@@ -0,0 +1,44 @@
+     * Return instance of the cluster context, must be initialized first for proper usage
+     * @return instance of cluster context
+     */
+    public static synchronized NeuralSearchSettingsAccessor instance() {
```
Sure, I can switch most of the singletons to injection. `PipelineServiceUtil`, `NeuralSearchSettingsAccessor`, and `StateStatsManager should be trivial. 

For the `EventStatsManager`, though, incrementing stats via static method call seems cleaner otherwise we'd have to add the instance to every single constructor of every object that wants to call it which is pretty invasive. I see OS core has an injector interface and uses @Inject for things like Transport Action, but it doesn't seem utilized in many other places, maybe for event stats its easier to keep the singleton instance

heemin32 replied:
Yes. For `EventStatsManager`, it might not suitable to inject.

martin-gaievski commented on src/test/java/org/opensearch/neuralsearch/rest/RestNeuralStatsHandlerIT.java:
```diff
@@ -0,0 +1,169 @@
+import java.util.Map;
+
+@Log4j2
+public class RestNeuralStatsHandlerIT extends BaseNeuralSearchIT {
```
we also need a BWC test. It's fine if there is no 2.x version that supports stats, it's needed for future 3.x releases

q-andy commented on src/main/java/org/opensearch/neuralsearch/settings/NeuralSearchSettingsAccessor.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public void initialize(ClusterService clusterService, Settings settings) {
+        if (initialized) return;
```
ack, this method was removed

q-andy reviewed (commented):
Addressed review comments, I'll change the format of the response according to the discussion above (breaking down high level categories as unflat like `all_nodes`) and add examples in the next commit.

heemin32 commented on src/main/java/org/opensearch/neuralsearch/stats/info/InfoStatsManager.java:
```diff
@@ -0,0 +1,195 @@
+
+    private void incrementBy(Map<InfoStatName, CountableInfoStatSnapshot> stats, InfoStatName statName, Long amount) {
+        if (stats.containsKey(statName)) {
+            stats.get(statName).incrementBy(amount);
```
Can we avoid `.get()` here as well?

q-andy commented:
Added BWC tests. Currently they don't run since there's no backwards versions, but included them to have basis for future 3.x versions.

Updated response formatting:

**Default**
```
GET {{ _.base_url }}/_plugins/_neural/stats
{
	"_nodes": {
		"total": 1,
		"successful": 1,
		"failed": 0
	},
	"cluster_name": "integTest",
	"info": {
		"cluster_version": "3.0.0",
		"processors": {
			"ingest": {
				"text_embedding_processors_in_pipelines": 0
			}
		}
	},
	"all_nodes": {
		"processors": {
			"ingest": {
				"text_embedding_executions": 0
			}
		}
	},
	"nodes": {
		"Qs3osnyfTz6AokNiPB2uRQ": {
			"processors": {
				"ingest": {
					"text_embedding_executions": 0
				}
			}
		}
	}
}
```
**Flatten**
```
GET {{ _.base_url }}/_plugins/_neural/stats?flat_stat_paths=true
{
	"_nodes": {
		"total": 1,
		"successful": 1,
		"failed": 0
	},
	"cluster_name": "integTest",
	"info": {
		"cluster_version": "3.0.0",
		"processors.ingest.text_embedding_processors_in_pipelines": 0
	},
	"all_nodes": {
		"processors.ingest.text_embedding_executions": 0
	},
	"nodes": {
		"Qs3osnyfTz6AokNiPB2uRQ": {
			"processors.ingest.text_embedding_executions": 0
		}
	}
}
```
**Flatten w/ metadata**
```
GET {{ _.base_url }}/_plugins/_neural/stats?flat_stat_paths=true&include_metadata=true
{
	"_nodes": {
		"total": 1,
		"successful": 1,
		"failed": 0
	},
	"cluster_name": "integTest",
	"info": {
		"cluster_version": {
			"value": "3.0.0",
			"stat_type": "settable_string"
		},
		"processors.ingest.text_embedding_processors_in_pipelines": {
			"value": 0,
			"stat_type": "info_countable"
		}
	},
	"all_nodes": {
		"processors.ingest.text_embedding_executions": {
			"value": 0,
			"stat_type": "timestamped_counter",
			"trailing_interval_value": 0,
			"minutes_since_last_event": 29028938
		}
	},
	"nodes": {
		"Qs3osnyfTz6AokNiPB2uRQ": {
			"processors.ingest.text_embedding_executions": {
				"value": 0,
				"stat_type": "timestamped_counter",
				"trailing_interval_value": 0,
				"minutes_since_last_event": 29028938
			}
		}
	}
}
```

martin-gaievski commented on qa/build.gradle:
```diff
@@ -45,7 +45,9 @@ dependencies {
         exclude group: 'org.slf4j', module: 'slf4j-api'
+        exclude group: 'net.minidev', module: 'json-smart'
     }
+    testRuntimeOnly group: 'net.minidev', name:'json-smart', version: "${versions.json_smart}"
```
you need to rebase on latest main and remove this

q-andy replied:
My [fork branch](https://github.com/q-andy/neural-search-os/tree/neural-stats) is up to date with main with latest commits, I think this diff is from the [feature branch](https://github.com/opensearch-project/neural-search/tree/feature/neural-stats-api.) on this repo being out of date. I don't have write permissions to sync it, could you try syncing it and this should disappear?

martin-gaievski reviewed (approved):
Looks good to me, thanks Andy. please address my comment regarding the rebasing on latest main

vibrantvarun reviewed (approved):
Looks good

vibrantvarun commented:
@heemin32 if your all comments are resolved then can we merge this?

heemin32 commented:
> @heemin32 if your all comments are resolved then can we merge this?

I think my comments are all resolved. G2G.
